src/DataProtection/DataProtection/src/Cng/CbcAuthenticatedEncryptor.cs

```diff
@@ -56,222 +57,162 @@ public CbcAuthenticatedEncryptor(Secret keyDerivationKey, BCryptAlgorithmHandle
+                    if (!ValidateHash(hashHandle, pbIV, _symmetricAlgorithmBlockSizeInBytes + (uint)cbEncryptedDataLength, pbActualHmac))
                     {
-                        DoCbcEncrypt(
-                            symmetricKeyHandle: symmetricKeyHandle,
-                            pbIV: pbIV,
-                            pbInput: &dummy,
-                            cbInput: 0,
-                            pbOutput: ptr,
-                            cbOutput: _symmetricAlgorithmBlockSizeInBytes);
+                        throw Error.CryptCommon_PayloadInvalid();
                     }
                 }
-                ptr += _symmetricAlgorithmBlockSizeInBytes;
 
-                // MAC a zero-length input string and copy the digest to the return buffer.
-                using (var hashHandle = _hmacAlgorithmHandle.CreateHmac(pbHmacSubkey, _hmacAlgorithmSubkeyLengthInBytes))
+                // If the integrity check succeeded, decrypt the payload.
+                using (var decryptionSubkeyHandle = _symmetricAlgorithmHandle.GenerateSymmetricKey(pbSymmetricEncryptionSubkey, _symmetricAlgorithmSubkeyLengthInBytes))
                 {
-                    hashHandle.HashData(
-                        pbInput: &dummy,
-                        cbInput: 0,
-                        pbHashDigest: ptr,
-                        cbHashDigest: _hmacAlgorithmDigestLengthInBytes);
+                    // BCryptDecrypt mutates the provided IV; we need to clone it to prevent mutation of the original value
+                    byte* pbClonedIV = stackalloc byte[checked((int)_symmetricAlgorithmBlockSizeInBytes)];
+                    UnsafeBufferUtil.BlockCopy(from: pbIV, to: pbClonedIV, byteCount: _symmetricAlgorithmBlockSizeInBytes);
+
+                    // First, query the output size needed
+                    var ntstatus = UnsafeNativeMethods.BCryptDecrypt(
+                        hKey: decryptionSubkeyHandle,
+                        pbInput: pbEncryptedData,
+                        cbInput: (uint)cbEncryptedDataLength,
+                        pPaddingInfo: null,
+                        pbIV: pbClonedIV,
+                        cbIV: _symmetricAlgorithmBlockSizeInBytes,
+                        pbOutput: null,  // NULL output = size query only
+                        cbOutput: 0,
+                        pcbResult: out var dwRequiredSize,
+                        dwFlags: BCryptEncryptFlags.BCRYPT_BLOCK_PADDING);
+                    UnsafeNativeMethods.ThrowExceptionForBCryptStatus(ntstatus);
+
+                    // Get buffer from writer with the required size
+                    var buffer = destination.GetSpan(checked((int)dwRequiredSize));
+
+                    // Clone IV again for the actual decryption call
+                    byte* pbClonedIV2 = stackalloc byte[checked((int)_symmetricAlgorithmBlockSizeInBytes)];
```

Comment:
This is the second call to `checked((int)_symmetricAlgorithmBlockSizeInBytes)`. While checked operations aren't *that* expensive, it's probably good to reuse the operation or just do the unchecked cast.

```suggestion
                    byte* pbClonedIV2 = stackalloc byte[(int)_symmetricAlgorithmBlockSizeInBytes];
```